sansio_lsp_client/client.py

```diff
@@ -171,31 +360,63 @@ def parse_request(event_cls: t.Type[Event]) -> Event:
+        if request.method == "workspace/workspaceFolders":
+            return parse_request(WorkspaceFolders)
+
+        elif request.method == "workspace/configuration":
+            return parse_request(ConfigurationRequest)
+
+        elif request.method == "window/showMessage":
             return parse_request(ShowMessage)
         elif request.method == "window/showMessageRequest":
             return parse_request(ShowMessageRequest)
         elif request.method == "window/logMessage":
             return parse_request(LogMessage)
+
         elif request.method == "textDocument/publishDiagnostics":
             return parse_request(PublishDiagnostics)
-        else:
-            raise NotImplementedError(request)
 
-    def recv(self, data: bytes) -> t.List[Event]:
+        elif request.method == "window/workDoneProgress/create":
+            ev = parse_request(WorkDoneProgressCreate)
+            self._progress_tokens_map[request.params['token']] = WorkDoneProgress
+            return parse_request(WorkDoneProgressCreate)
+
+        elif request.method == "$/progress":
+            progress_type = self._progress_tokens_map.get(request.params['token'])
+
+            if progress_type == WorkDoneProgress:
+                kind = request.params.get('value', {}).get('kind')
+                if kind == 'begin':
+                    return parse_request(WorkDoneProgressBegin)
+                elif kind == 'report':
+                    return parse_request(WorkDoneProgressReport)
+                elif kind == 'end':
+                    del self._progress_tokens_map[request.params["token"]]
+                    return parse_request(WorkDoneProgressEnd)
+
+        elif request.method == "client/registerCapability":
+            return parse_request(RegisterCapabilityRequest)
+
+        raise NotImplementedError(request)
+
+    def recv(self, data: bytes, errors: t.Optional[list] = None) -> t.List[Event]:
         self._recv_buf += data
 
         # _parse_messages deletes consumed data from self._recv_buf
         messages = list(_parse_messages(self._recv_buf))
 
         events: t.List[Event] = []
         for message in messages:
-            if isinstance(message, Response):
-                events.append(self._handle_response(message))
-            elif isinstance(message, Request):
-                events.append(self._handle_request(message))
-            else:
-                raise RuntimeError("nobody will ever see this, i hope")
+            try:
+                if isinstance(message, Response):
+                    events.append(self._handle_response(message))
+                elif isinstance(message, Request):
+                    events.append(self._handle_request(message))
+                else:
+                    raise RuntimeError("nobody will ever see this, i hope")
+            except Exception as ex:
+                if errors is not None:
+                    errors.append(ex)
```

Comment:
If `errors` is `None`, this swallows the exception; it doesn't display it or store it somewhere for later. Do we really need an `errors` list like this?

Reply:
Yeah I'm not a fan of the current design of `errors`. I could see the value in having some way to handle errors and keep going, but having an errors list passed in as a parameter is not the way to go for me.

Reply:
This is meant for `NotImplementedError`, so it's ok for normal user to not care about it. I'll change `catch` to this specific exception, and the other ones can roam free, deal? :)

Reply:
What problem does the `errors` list solve? What was throwing `NotImplementedError`? What did you do to invoke it?

Reply:
1. In `data` argument of `recv()` there can be several messages, so for one broken message to not affect (following) proper messages I added `try` block.
2. There can be unimportant stuff. For example C# server (omnisharp) sends these:
    ```
    ... "method":"o#/msbuildprojectdiagnostics" ...
    ... "method":"o#/projectadded" ...
    ... "method":"o#/projectchanged" ...
    ```
    And probably important errors... so I added a simple way to get exceptions that occurred. This is the best I could come up for now. I'm open to suggestions here.

Reply:
So can we just use `logging` for all the unknown stuff?

Reply:
For unknown `methods`? Great idea. And `logger.error()` seems like an ok place for all the other exceptions here.

Reply:
I usually use `logger.exception`, which logs with severity `ERROR` and also includes the traceback.